File: RideSharingSystem.java
import java.util.concurrent.*;
import java.util.*;
import java.io.*;

// Author: Nabin
public class RideSharingSystem {

    // Shared task queue
    private final BlockingQueue<String> taskQueue = new LinkedBlockingQueue<>();

    // Shared result list
    private final List<String> resultList = Collections.synchronizedList(new ArrayList<>());

    // Logger to track system events
    private final PrintWriter logger;

    public RideSharingSystem() throws IOException {
        logger = new PrintWriter(new FileWriter("log_java.txt"), true);
    }

    // Adds tasks to the shared queue
    public void addTask(String task) {
        taskQueue.offer(task);
    }

    // Worker class implementing Runnable
    class Worker implements Runnable {
        private final int id;

        public Worker(int id) {
            this.id = id;
        }

        @Override
        public void run() {
            logger.println("Thread " + id + " started.");
            while (true) {
                try {
                    String task = taskQueue.poll(1, TimeUnit.SECONDS);
                    if (task == null) break; // Exit when no tasks are left
                    // Simulate processing
                    Thread.sleep(500);
                    String result = "Thread " + id + " processed: " + task;
                    resultList.add(result);
                    logger.println(result);
                } catch (InterruptedException e) {
                    logger.println("Thread " + id + " interrupted.");
                    Thread.currentThread().interrupt();
                } catch (Exception e) {
                    logger.println("Thread " + id + " error: " + e.getMessage());
                }
            }
            logger.println("Thread " + id + " finished.");
        }
    }

    public void execute(int numberOfThreads) {
        ExecutorService executor = Executors.newFixedThreadPool(numberOfThreads);
        for (int i = 0; i < numberOfThreads; i++) {
            executor.execute(new Worker(i));
        }
        executor.shutdown();
        try {
            executor.awaitTermination(5, TimeUnit.MINUTES);
        } catch (InterruptedException e) {
            logger.println("Executor interrupted: " + e.getMessage());
        }

        // Save results
        try (PrintWriter out = new PrintWriter("results_java.txt")) {
            for (String res : resultList) {
                out.println(res);
            }
        } catch (IOException e) {
            logger.println("Failed to save results: " + e.getMessage());
        }
    }

    public static void main(String[] args) {
        try {
            RideSharingSystem system = new RideSharingSystem();

            // Add tasks to the queue
            for (int i = 1; i <= 20; i++) {
                system.addTask("Ride Request " + i);
            }

            // Start system with 4 worker threads
            system.execute(4);

        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: ride_sharing_system.go
package main

import (
	"fmt"
	"log"
	"os"
	"sync"
	"time"
)

// Author: Nabin

var (
	taskQueue   = make(chan string, 20)
	resultList  = []string{}
	mutex       sync.Mutex
	logFile     *os.File
	waitGroup   sync.WaitGroup
)

func init() {
	var err error
	logFile, err = os.OpenFile("log_go.txt", os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
	if err != nil {
		log.Fatal("Cannot open log file:", err)
	}
	log.SetOutput(logFile)
}

// Worker function simulates ride processing
func worker(id int) {
	defer waitGroup.Done()
	log.Printf("Goroutine %d started.\n", id)
	for task := range taskQueue {
		time.Sleep(500 * time.Millisecond) // simulate processing
		result := fmt.Sprintf("Goroutine %d processed: %s", id, task)

		// Safely append result
		mutex.Lock()
		resultList = append(resultList, result)
		mutex.Unlock()

		log.Println(result)
	}
	log.Printf("Goroutine %d finished.\n", id)
}

func main() {
	defer logFile.Close()

	// Add tasks to queue
	for i := 1; i <= 20; i++ {
		taskQueue <- fmt.Sprintf("Ride Request %d", i)
	}
	close(taskQueue)

	// Start 4 workers
	for i := 0; i < 4; i++ {
		waitGroup.Add(1)
		go worker(i)
	}

	waitGroup.Wait()

	// Save results to file
	resultFile, err := os.Create("results_go.txt")
	if err != nil {
		log.Println("Error creating result file:", err)
		return
	}
	defer resultFile.Close()

	for _, res := range resultList {
		fmt.Fprintln(resultFile, res)
	}
}
